Add an Informer example in client-go

wallrj opened the pull request:
This PR adds an example demonstrating how to use the client-go library to set up an Informer to print out Pod events and to maintain a local store containing Pod configuration synchronized with the Kubernetes API server.

I've copied the work started in https://github.com/kubernetes/client-go/pull/30 and addressed most of the outstanding code review comments:
https://github.com/kubernetes/client-go/pull/30

```release-note
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44446)
<!-- Reviewable:end -->

wallrj commented:
I'm struggling to sign the CLA as an employee of jetstack.io.
The branch originally contained commits with my generic email github project email address `wallrj@users.noreply.github.com` but I've force pushed a new single commit using my `jetstack.io` address.
(I'm told that jetstack.io have completed the corporate CLA process for CNCF.)

grodrigues3 commented:
@k8s-bot ok to test

grodrigues3 commented:
bump the CLA check

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,146 @@
+	"k8s.io/client-go/tools/clientcmd"
+
+	// Only required to authenticate against GKE clusters
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
```
Can we keep this line but comment it out? This line recursively introduces a huge dependency.

wallrj replied:
Done.

caesarxuchao commented:
FWIW, https://github.com/kubernetes/kubernetes/pull/44320 adds an example showing how to use workqueue together with informer. Its emphasis is on the use of the workqueue.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,146 @@
+	s := <-signals
+	fmt.Printf("received signal %#v, exiting...\n", s)
+	close(stop)
+	os.Exit(0)
```
Curious why os.Exit(0) is needed here?

wallrj replied:
I think the intent was to exit with status 0 after handling SIGINT.
Default behaviour is to exit with 1.
I agree though, it distracts from the Informer code.
Removed.

ash2k commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,146 @@
+	source := cache.NewListWatchFromClient(
+		clientset.Core().RESTClient(),
+		"pods",
+		api.NamespaceAll,
```
Should it be a versioned package like meta/v1 or api/v1? They also have this constant.

wallrj replied:
Done.

ash2k commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,146 @@
+		panic(err)
+	}
+
+	stop := make(chan struct{}, 1)
```
Doesn't have to be buffered?

wallrj replied:
Agreed. Done.

rmohr commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,137 @@
+	return fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+}
+
+func buildConfig(kubeconfig string) (*rest.Config, error) {
```
This helper method is not needed, `BuildConfigFromFlags`, checks if a `masterUrl` or a `kubeconfigPath` is set. If both are unset, it assumes that it is running from inside the cluster.

rmohr commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,137 @@
+	// Create the client config. Use kubeconfig if given, otherwise assume in-cluster.
+	config, err := buildConfig(*kubeconfig)
+	if err != nil {
+		panic(err)
```
You can use `glog.Fatal(err)`

rmohr commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,137 @@
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err)
```
You can use `glog.Fatal(err)`

kargakis commented:
@mohr already added an example [here](https://github.com/kubernetes/kubernetes/pull/44320)

caesarxuchao commented:
@kargakis the one you linked to emphasis more on the workqueue and is more complex. I think it doesn't hurt to have a simpler example of how to use informer alone. WDYT?

caesarxuchao commented:
@wallrj a few commands you need to run to pass the verification script:
```
W0418 01:42:06.769] 1 BUILD files not up-to-date.
I0418 01:42:06.870] 
I0418 01:42:06.870] Run ./hack/update-bazel.sh
```

```
I0418 01:53:26.662] FAILED   hack/make-rules/../../hack/verify-golint.sh	37s
I0418 01:53:26.662] Skipping hack/make-rules/../../hack/verify-govet.sh
I0418 01:53:26.662] Verifying hack/make-rules/../../hack/verify-import-boss.sh
W0418 01:53:26.763] Some packages passed golint but are not listed in hack/.linted_packages.
W0418 01:53:26.763] Please add them in alphabetical order:
W0418 01:53:26.763] 
W0418 01:53:26.764]   echo staging/src/k8s.io/client-go/examples/informer >> hack/.linted_packages
```

Also, the CLA check is also failing, have you signed it?

wallrj commented:
Sorry for the delay. And thanks for the review @rmohr and @caesarxuchao.
I have now signed the CLA and I'll address your comments this week.

wallrj commented:
I've addressed all code review feedback and the tests seem to pass.
Not sure whether I need to add release notes.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,134 @@
+
+	// store can be used to List and Get
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	fmt.Println("listing pods from store:")
```
Since you are importing glog already, might just use glog.Info() here as well.

caesarxuchao commented:
Thanks @wallrj. A nit and please squash, then lgtm.

wallrj commented:
@k8s-bot bazel test this

wallrj commented:
@caesarxuchao 
 * Used glog throughout.
 * Added some instructions on running the example InCluster.
 * Squashed the commits.
 * Tested (manually) with Minikube.

Let me know if there's anything else to do.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/README.md:
````diff
@@ -0,0 +1,86 @@
+```
+
+## Use Cases
+
````
The main usage of informer is writing controllers. I think it is worth adding something like the following in `use cases`:
* Building controllers that coordinate other resources. Most controllers in [k8s.io/kubernetes/pkg/controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller) use informers.

wallrj replied:
Done.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/README.md:
```diff
@@ -0,0 +1,86 @@
+
+* How to write an Informer against a core resource type.
+* How to handle add, update and delete events.
+
```
Could you link to the thirdparty resource example? Something like "The [thirdparty-resources](../third-party-resources) example demonstrates how to use informers against a thirdparty resource"

wallrj replied:
Done.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/README.md:
```diff
@@ -0,0 +1,86 @@
+
+* Capturing resource events for logging to external systems
+  (e.g. monitor non-"Normal" events and publish metrics to a time series database)
+* Creating lifecycle controllers for ThirdPartyResources
```
I think this is included in the "Building controllers" use case i suggested in my comment, consider remove this one.

wallrj replied:
Done.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,136 @@
+		glog.Info(pod.ObjectMeta.Name)
+	}
+
+	select {}
```
What is this line for?

wallrj replied:
It blocks the main thread of execution so the goroutines have
time to run.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,136 @@
+
+// Handler functions as per the controller above.
+// Note the coercion of the interface{} into a pointer of the expected type.
+
```
Remove the extra line?

wallrj replied:
Done.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+// convenience functions
+
```
Remove the extra line?

wallrj replied:
Done.

caesarxuchao reviewed (changes requested):
Sorry i have a few more requests.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/README.md:
```diff
@@ -0,0 +1,87 @@
+
+## Use Cases
+
+* Building controllers that coordinate other resources. Most controllers in k8s.io/kubernetes/pkg/controller use informers.
```
hyperlink `k8s.io/kubernetes/pkg/controller`? https://github.com/kubernetes/kubernetes/tree/master/pkg/controller

wallrj replied:
Done. Hyperlinked to godocs instead of code: https://godoc.org/k8s.io/kubernetes/pkg/controller

caesarxuchao commented on staging/src/k8s.io/client-go/examples/informer/README.md:
```diff
@@ -0,0 +1,87 @@
+
+## Running Inside a Kubernetes Cluster
+
+You can also run the example inside a Kubernetes cluster.
```
Could we omit the instruction on how to start minikube? After all there are many other ways to get a kubernetes cluster. Instead just something like "Check out this [guide](https://kubernetes.io/docs/setup/) on how to setup a Kubernetes cluster".

wallrj replied:
Done.

caesarxuchao reviewed (commented):
Two more nits :)

wallrj committed:
Add an Informer example in client-go

wallrj committed:
Hopefully this version is clearer.

wallrj committed:
oops, I'd disabled the resync

wallrj commented:
@caesarxuchao Thanks for the continuing reviews :-)

I addressed your comments and pushed an update.

Then I took a closer look at e.g. the Kubernetes Deployment controller and found it to much closer to @rmohr 's work queue example. 

And I went looking at the client-go/informers package and found generated packages that provide a much clearer API....this version 

So I've updated the example and I think it's now much cleaner and has less overlap with the work queue example.

See what you think.

wallrj committed:
update-bazel

krmayankk commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,97 @@
+	// wait until a store finished its initial synchronization
+	cache.WaitForCacheSync(stop, informer.HasSynced)
+
+	informer.AddEventHandler(
```
do we need to add the event handlers before we start the Run loop ?

wallrj replied:
I deliberately added them *after* synchronisation, intending that it wouldn't log all the existing Pods but it seems to log them all anyway.

```
[richard@drax informer]$ go run main.go -kubeconfig=$HOME/.kube/config -logtostderr
I0525 09:43:27.434950   20400 main.go:76] POD CREATED: kube-system/kube-addon-manager-minikube
I0525 09:43:27.435084   20400 main.go:76] POD CREATED: kube-system/kube-dns-268032401-sf2pp
I0525 09:43:27.435101   20400 main.go:76] POD CREATED: kube-system/kubernetes-dashboard-69vl2

```

That's probably worth a comment.

ncdc replied:
See my comment above

krmayankk commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,97 @@
+
+	factory := informers.NewSharedInformerFactory(clientset, time.Second*10)
+
+	informer := factory.Core().V1().Pods().Informer()
```
it might be helpful to put some comments on whether this requires a informer per k8s resource. If i need to watch both pod and services for e.g. do i need to create two informers ? If there is a doc about the informer framework , please point to me ?

krmayankk replied:
@wallrj also this might be asking too much, but do you plan to create a similar example with the sharedInformer framework which is now being used by a lot of the controllers ?

wallrj replied:
> it might be helpful to put some comments on whether this requires a informer per k8s resource.

Yes, you need to create an informer for each resource. 
For example, the Deployment controller uses informers for Deployment, ReplicaSet and Pod:
 * https://github.com/kubernetes/kubernetes/blob/v1.6.4/pkg/controller/deployment/deployment_controller.go#L119

>  If there is a doc about the informer framework , please point to me ?


I don't know of any up-to-date documentation for Informer....perhaps this is it!

> @wallrj also this might be asking too much, but do you plan to create a similar example with the sharedInformer framework which is now being used by a lot of the controllers ?

This example uses a Shared Informer. See the line above:

`factory := informers.NewSharedInformerFactory(...`

krmayankk replied:
adding @smarterclayton for more suggestions here. Is the queue per resource type( and since informer is per resource type) and also  the queue is per informer here ?

ncdc replied:
@krmayankk what "queue" are you referring to? For the shared informers that we autogenerate via code generation, there is one shared informer per resource type.

caesarxuchao commented:
@wallrj if showing how to use sharedInformers (which i agree is very valuable), i think the example will be more useful if the informers are actually used by multiple controllers. And when used by multiple controllers, there are many non-obvious things worth documenting, like what i [learned](https://github.com/kubernetes/kubernetes/pull/45427#discussion_r117273724) last week :)

I'm not familiar with sharedInformers myself. @ncdc @deads2k do you have suggestions on what should be included in a sharedInformer's example?

ncdc commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,97 @@
+	defer close(stop)
+
+	// the controller run starts the event processing loop
+	go informer.Run(stop)
```
Don't call `informer.Run()`. You'll want to call `factory.Start()`, _but_ the overall order needs to be:

1. Create shared informer factory
1. Add event handler to informer you get from factory
1. Start shared informer factory
1. Wait for cache sync

With a controller, because it normally runs in its own separate goroutine, you can add the event handler and the immediately wait for the cache sync, because we know at some point later on in the main initialization code, the shared informer factory will be started at the very end of the process.

wallrj replied:
Done.

ncdc commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,97 @@
+		glog.Fatal(err)
+	}
+
+	factory := informers.NewSharedInformerFactory(clientset, time.Second*10)
```
10 seconds is a pretty small resync period. You'd probably want it to be at least 5 minutes, although somewhere between 1 and 24 hours might be best. The resync period determines how often the cache is reprocessed, and when this happens, every item in the cache is redelivered to the event handlers.

wallrj replied:
Done.

ncdc commented:
If you are using a shared informer, you may not mutate the object you receive in the event handler. If you need to modify it, you must first make a deep copy. We have a cache mutation detector that you can enable (set the environment variable `KUBE_CACHE_MUTATION_DETECTOR` to true) to detect invalid mutations. Note that this check is cpu intensive and should only be used when running tests (unit, integration, e2e) and not for real deployments.

krmayankk commented on staging/src/k8s.io/client-go/examples/informer/main.go:
```diff
@@ -0,0 +1,97 @@
+				newPod := new.(*v1.Pod)
+				glog.Infof(
+					"POD UPDATED. %s/%s %s",
+					oldPod.Namespace, oldPod.Name, newPod.Status.Phase,
```
@wallrj also may be add a comment saying the objects in the event handlers cannot be mutated. If we want to modify or update those objects, we need to clone them , etc

krmayankk replied:
@wallrj later maybe we should add an exmaple with two informers(later PR)

wallrj committed:
Use a longer resync period

wallrj committed:
Use factory.Start

wallrj commented:
Thanks again for the continuing reviews.

I'm going to try and add some unit tests for this code.
I'm interested to know which of the Kubernetes Fakes you'd use.
And how you'd simulate resource changes.

wallrj committed:
Make the example more like a Kubernetes controller and add some tests.

wallrj committed:
./hack/update-bazel.sh

wallrj commented:
I've updated the example to be more like a Kubernetes controller.
But I'd like advice on how one might test it.
The `fake.NewSimpleClient` used in kubernetes tests doesn't seem to trigger Add, Update, Delete event handlers.
Do people not test at that level?

wallrj committed:
Add missing boilerplate

wallrj committed:
Document public functions

wallrj commented:
`NewFakeControllerSource` seems like it'd be perfect for my tests.
It'd allow me to add objects and trigger the `Add / Update / Delete` events but I can't find any place to inject it when using SharedInformers and SharedInformerFactory.

@caesarxuchao, @ncdc Any ideas?

ncdc commented:
@wallrj I added a line comment showing you how to use the fake watch to get the event handler functions to fire. Let me know if you need more help.

krmayankk commented:
@ncdc @caesarxuchao i wouldn't block on a great example if this would help a lot of people using the informer. This is an example code. i am guessing @wallrj would loose motivation at some point :-)

caesarxuchao commented:
Sorry i'm working on 1.7 release items. I'll try to take another look this week.

ahmetb commented on staging/src/k8s.io/client-go/examples/informer/README.md:
````diff
@@ -0,0 +1,34 @@
+```sh
+go run main.go -kubeconfig=$HOME/.kube/config -logtostderr
+```
+
````
Can you explain what happens when someone runs this program? Example of this is at: https://github.com/kubernetes/client-go/tree/master/examples/create-update-delete-deployment We're trying to keep READMEs of all the client examples similar to each other.

ahmetb commented on staging/src/k8s.io/client-go/examples/informer/README.md:
```diff
@@ -0,0 +1,34 @@
+* Building controllers that coordinate other resources.
+  Most controllers in [k8s.io/kubernetes/pkg/controller](https://godoc.org/k8s.io/kubernetes/pkg/controller) use informers.
+* Capturing resource events for logging to external systems
+  (e.g. monitor non-"Normal" events and publish metrics to a time series database)
```
Can you add a `## Cleanup` section people can run to clean up the artifacts created by running this example? (For instance: https://github.com/kubernetes/client-go/tree/master/examples/create-update-delete-deployment or https://github.com/kubernetes/client-go/tree/master/examples/in-cluster has this section). 

This will help us keep the README files for each client-go example similar.

wallrj committed:
simulate pods being added

wallrj committed:
Use new k8s/io/api repo

wallrj committed:
update-bazel.sh

wallrj committed:
Never resync

wallrj commented on staging/src/k8s.io/client-go/examples/informer/main_test.go:
```diff
@@ -0,0 +1,112 @@
+				clientset := fake.NewSimpleClientset(tc.initial...)
+				factory := informers.NewSharedInformerFactory(
+					clientset,
+					time.Hour*24,
```
Done.

wallrj committed:
Experimenting with fuzz testing.

wallrj committed:
Restrict the names and namespaces and perform delete operations too
